File: Frameworks/flutter-local/test/regression/example_specs_test.dart
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:ods_flutter_local/parser/spec_parser.dart';

/// Regression test: every example spec in the Specification repo must parse
/// successfully with no validation errors.
void main() {
  final parser = SpecParser();
  final examplesDir = Directory('../../Specification/Examples');

  // Skip gracefully if running in CI without the Specification repo.
  if (!examplesDir.existsSync()) {
    test('SKIP: Specification/Examples not found', () {
      // Not a failure — the Specification repo may not be adjacent in CI.
    });
    return;
  }

  final specFiles = examplesDir
      .listSync()
      .whereType<File>()
      .where((f) => f.path.endsWith('.json') && !f.path.endsWith('catalog.json'))
      .toList()
    ..sort((a, b) => a.path.compareTo(b.path));

  group('Example specs parse without errors', () {
    for (final file in specFiles) {
      final name = file.uri.pathSegments.last;
      test(name, () {
        final json = file.readAsStringSync();
        final result = parser.parse(json);

        // Must parse successfully.
        expect(result.parseError, isNull,
            reason: 'Parse error in $name: ${result.parseError}');

        // Must have no validation errors (warnings are OK).
        expect(result.validation.hasErrors, false,
            reason: 'Validation errors in $name: ${result.validation.errors}');

        // Must produce a valid app model.
        expect(result.app, isNotNull, reason: '$name produced null app');
        expect(result.app!.appName, isNotEmpty,
            reason: '$name has empty appName');
        expect(result.app!.pages, isNotEmpty,
            reason: '$name has no pages');
      });
    }
  });

  group('Example spec structure checks', () {
    for (final file in specFiles) {
      final name = file.uri.pathSegments.last;
      test('$name: startPage exists in pages', () {
        final json = file.readAsStringSync();
        final result = parser.parse(json);
        if (result.app == null) return; // Covered by parse test above.

        expect(result.app!.pages.containsKey(result.app!.startPage), true,
            reason:
                '$name: startPage "${result.app!.startPage}" not found in pages');
      });
    }
  });
}

File: Frameworks/react-web/tests/unit/regression/example-specs.test.ts
import { describe, it, expect } from 'vitest'
import fs from 'fs'
import path from 'path'
import { parseSpec, isOk } from '../../../src/parser/spec-parser.ts'

/// Regression test: every example spec in the Specification repo must parse
/// successfully with no validation errors.
/// Ported from flutter-local/test/regression/example_specs_test.dart.

const examplesDir = path.resolve(__dirname, '../../../../../Specification/Examples')

// Files that are not app specs (e.g., catalog/index files).
const SKIP_FILES = new Set(['catalog.json'])

function getSpecFiles(): string[] {
  if (!fs.existsSync(examplesDir)) {
    return []
  }
  return fs
    .readdirSync(examplesDir)
    .filter((f) => f.endsWith('.json') && !SKIP_FILES.has(f))
    .sort()
}

const specFiles = getSpecFiles()

if (specFiles.length === 0) {
  describe('Example specs (SKIPPED)', () => {
    it('SKIP: Specification/Examples not found', () => {
      // Not a failure — the Specification repo may not be adjacent in CI.
    })
  })
} else {
  describe('Example specs parse without errors', () => {
    for (const fileName of specFiles) {
      it(fileName, () => {
        const filePath = path.join(examplesDir, fileName)
        const json = fs.readFileSync(filePath, 'utf-8')
        const result = parseSpec(json)

        // Must parse successfully.
        expect(result.parseError, `Parse error in ${fileName}: ${result.parseError}`).toBeNull()

        // Must have no validation errors (warnings are OK).
        expect(
          result.validation.hasErrors,
          `Validation errors in ${fileName}: ${result.validation.errors.map((e) => e.message).join(', ')}`,
        ).toBe(false)

        // Must produce a valid app model.
        expect(result.app, `${fileName} produced null app`).not.toBeNull()
        expect(result.app!.appName, `${fileName} has empty appName`).toBeTruthy()
        expect(Object.keys(result.app!.pages).length, `${fileName} has no pages`).toBeGreaterThan(0)
      })
    }
  })

  describe('Example spec structure checks', () => {
    for (const fileName of specFiles) {
      it(`${fileName}: startPage exists in pages`, () => {
        const filePath = path.join(examplesDir, fileName)
        const json = fs.readFileSync(filePath, 'utf-8')
        const result = parseSpec(json)
        if (result.app == null) return // Covered by parse test above.

        expect(
          result.app.startPage in result.app.pages,
          `${fileName}: startPage "${result.app.startPage}" not found in pages`,
        ).toBe(true)
      })
    }
  })
}
